README.md

```diff
@@ -1,3 +1,607 @@
+        signature
+      </code>
+    </th>
+    <td>
+      String
+    </td>
+    <td>
+      <strong>Required.</strong>
+      Signature string as calculated by <code><a href="#sign">sign()</a></code>.
+    </td>
+  </tr>
+</table>
+
+### Receiver
+
+```js
+const createReceiver = require('@octokit/webhooks').receiver
+// or: const createReceiver = require('@octokit/webhooks/receiver')
+
+const receiver = createReceiver({secret: 'mysecret'})
+
+receiver.on('installation', handleInstallationEvent)
+receiver.hook('installation', asyncInstallationHook)
+
+// assuming request is an object with .headers & .body properties
+receiver.handle({
+  name: request.headers['X-GitHub-Event'],
+  data: request.body,
+  signature: request.headers['X-Hub-Signature']
+}).catch(handleErrorsFromHooks)
+```
+
+`createReceiver({secret})` returns a `receiver` instance. The `secret` option is _not_ required.
+
+<a name="receiver-handle"></a>
+
+The `receiver` has a `.handle({name, data, signature})` method which returns a Promise and expects the following options
+
+<table width="100%">
+  <tr>
+    <th>
+      <code>
+        name
+      </code>
+    </th>
+    <td>
+      String
+    </td>
+    <td>
+      <strong>Required.</strong>
+      Name of the event. Passed as <a href="https://developer.github.com/webhooks/#delivery-headers"><code>X-GitHub-Event</code> header</a>
+      in the webhook request.
+    </td>
+  </tr>
+  <tr>
+    <th>
+      <code>
+        data
+      </code>
+    </th>
+    <td>
+      Object
+    </td>
+    <td>
+      <strong>Required.</strong>
+      Webhook request body as received from GitHub
+    </td>
+  </tr>
+  <tr>
+    <th>
+      <code>
+        signature
+      </code>
+    </th>
+    <td>
+      String
+    </td>
+    <td>
+      <strong>Required if <code>secret</code> passed to <code>createReceiver</code>.</strong>
+      Passed as <a href="https://developer.github.com/webhooks/#delivery-headers"><code>X-Hub-Signature</code> header</a>
+      in the webhook request.
+    </td>
+  </tr>
+</table>
+
+Besides the `.handle({name, data, signature})` method, the `receiver` instance also exposes the [Events & Hook](#events-and-hooks) methods.
+
+### Middleware
+
+```js
+const createMiddleware = require('@octokit/webhooks').middleware
+// or: const createMiddleware = require('@octokit/webhooks/middleware')
+
+const middleware = createMiddleware({ secret: 'mysecret' })
+
+middleware.on('installation', handleInstallationEvent)
+middleware.hook('installation', installationEventHook)
+
+http.createServer(function (request, response) {
+  if (request.url !== '/github-webhooks' || request.method !== 'POST') {
+    response.statusCode = 404
+    response.end('Not found')
+    return
+  }
+
+  middleware(request, response)
+}).listen(3000)
+```
+
+`createMiddleware({secret})` returns a `middleware` function. The `secret` option is _not_ required.
+
+The `middleware` function also exposes the [Events & Hook](#events-and-hooks) methods.
```

Comment:
I think we also need to expose the `.handle()` method from the middleware’s `receiver` instance. This make testing easier, and we currently expose a `.receive` function on Probot’s `robot` instance here: https://github.com/probot/probot/blob/83b7454f74160d81193e447012e3c507c0092c4e/lib/robot.js#L22-L26

I’m not sure if we should call the method `middleware.handle()` though, or something that makes more clear that it’s meant for testing or at least make folks think twice before using it, as it works around some safety measurements that the middleware implements, e.g. `middleware.inject()`. Any thoughts on that?

Reply:
Another possibility would be to support an API like this

```js
const Webhooks = require('@octokit/webhooks')
const webhooks = new Webhooks({secret: 'mysecret', middleware: {path: '/custom/webhooks'})

// webhooks has pre-initialised .middleware & .receiver. 
// webhooks.middleware is using the webhooks.reicever instance internally

const express = require('express')
const app = express()
app.use(options.webhookPath || '/', webhooks.middleware)

const createRobot = require('./lib/robot')
const robot = createRobot({app, webhooks})

// the createRobot method sets
// robot.on = webhooks.receiver.on
// robot.removeListener = webhooks.receiver.removeListener
// robot.inject = webhooks.receiver.handle
```

^^ This makes a lot of sense to me. Instead of exposing the same static methods at `require('@octokit/webhooks').middleware` and `require('@octokit/webhooks/middleware')`, I would get rid of `require('@octokit/webhooks').middleware`, make `Webhooks = require('@octokit/webhooks')` a factory or a class `.middleware` and the other modules and methods pre-initialised with the secret passed to `new Webhooks(options)`

That would let folks use e.g. `'@octokit/webhooks/middleware'` standalone. But if they want to inject an event or use the `.sign` / `.verify` method, they would use the `@octokit/webhooks` module which puts it all together.

I will give this some more thought and will try this out on probot. Happy to hear what y’all think.